src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/GenericVirtualMethodImplNode.cs

```diff
@@ -0,0 +1,78 @@
+        private readonly MethodDesc _method;
+
+        public GenericVirtualMethodImplNode(MethodDesc method)
+        {
+            Debug.Assert(method.GetCanonMethodTarget(CanonicalFormKind.Specific) == method);
+            Debug.Assert(method.HasInstantiation);
+
+            // This is either a generic virtual method or a MethodImpl for a static interface method.
+            // We can't test for static MethodImpl so at least sanity check it's static and noninterface.
+            Debug.Assert(method.IsVirtual || (method.Signature.IsStatic && !method.OwningType.IsInterface));
+
+            _method = method;
+        }
+
+        public override bool HasConditionalStaticDependencies => false;
+        public override bool InterestingForDynamicDependencyAnalysis => false;
+        public override bool StaticDependenciesAreComputed => true;
+        protected override string GetName(NodeFactory factory) => "__GVMImplNode_" + factory.NameMangler.GetMangledMethodName(_method);
+
+        public override IEnumerable<DependencyListEntry> GetStaticDependencies(NodeFactory factory)
+        {
+            DependencyList dependencies = null;
+
+            factory.MetadataManager.GetDependenciesDueToVirtualMethodReflectability(ref dependencies, factory, _method);
+
+            bool validInstantiation =
+                _method.IsSharedByGenericInstantiations || (      // Non-exact methods are always valid instantiations (always pass constraints check)
+                    _method.Instantiation.CheckValidInstantiationArguments() &&
+                    _method.OwningType.Instantiation.CheckValidInstantiationArguments() &&
+                    _method.CheckConstraints());
+
+            if (validInstantiation)
+            {
+                if (factory.TypeSystemContext.SupportsUniversalCanon && _method.IsGenericDepthGreaterThan(UniversalCanonGVMDepthHeuristic_CanonDepth))
+                {
+                    // fall back to using the universal generic variant of the generic method
+                    return dependencies;
+                }
+
+                bool getUnboxingStub = _method.OwningType.IsValueType && !_method.Signature.IsStatic;
+                dependencies ??= new DependencyList();
+                dependencies.Add(factory.MethodEntrypoint(_method, getUnboxingStub), "GVM Dependency - Canon method");
+
+                if (_method.IsSharedByGenericInstantiations)
+                {
+                    dependencies.Add(factory.NativeLayout.TemplateMethodEntry(_method), "GVM Dependency - Template entry");
+                    dependencies.Add(factory.NativeLayout.TemplateMethodLayout(_method), "GVM Dependency - Template");
+                }
+            }
+
+            return dependencies;
+        }
+
+        public override IEnumerable<CombinedDependencyListEntry> GetConditionalStaticDependencies(NodeFactory context) => null;
+
+        public override bool HasDynamicDependencies => false;
```

Comment:
It seems like this is important, but it's not clear to me why.

Reply:
All nodes have HasDynamicDependencies false, except for GVMDependencies node.

Returning true means the dependency analysis will call into `SearchDynamicDependencies` after every iteration of the graph expansion with a list of all nodes we've seen so far.

This node not reporting it is the optimization.

Reply:
And why is this a valid optimization? Or, why is it necessary for GVMDependencies to always return true?

Reply:
`GVMDependencies` needs to look at each type in the graph that has something to do with generic virtual methods to check whether the type provides an override of the slot. If so, it needs to instantiate the override with whatever generic method instantiation this `GVMDependencies` has.

But we don't need to do this for generic virtual method implementations that don't define a new slot (i.e. they're just overrides) - this is what the optimization tries to split - a slot use + an override ("implementation", since we also introduce this node for non abstract new slots) of the slot.